exercises/src/test/scala/datastructures/ListSpec.scala

```diff
@@ -0,0 +1,417 @@
+  //#3.8
+  "foldRight" should {
+    "result in original list when passed the case class constructors" in {
+      List.foldRight(List(), Nil:List[Int])(Cons(_,_)) mustEqual List()
+      List.foldRight(List(1,2,3), Nil:List[Int])(Cons(_,_)) mustEqual List(1,2,3)
+    }
+  }
+
+  //========================================================================================
+  //#3.9
+  "length" should {
+    "result in 0 for an empty List" in {
+      List.length(List()) mustEqual 0
+    }
+
+    "result in 1 for a single-element List" in {
+      List.length(List(1)) mustEqual 1
+      List.length(List('a')) mustEqual 1
+    }
+
+    "reslut in the number of elements in a multiple-element List" in {
+      List.length(List(1,2,3)) mustEqual 3
+      List.length(List('a','b','c')) mustEqual 3
+    }
+  }
+
+  // //#3.10
+  // "foldLeft" should {
+  //   "result in original list when passed the case class constructors" in {
+  //     List.foldLeft(List(), Nil:List[Int])((b,a) =>Cons(a,b)) mustEqual List()
+  //     List.foldLeft(List(1,2,3), Nil:List[Int])((b,a) =>Cons(a,b)) mustEqual List(3,2,1)
+  //   }
+  // }
+
+ //========================================================================================
+ //#3.11
+  "sum3 (uses foldLeft)" should {
+    "result in zero for an empty List" in {
+      List.sum3(List()) mustEqual 0
+    }
+
+    "return the element value in a single-element List" in {
+      List.sum3(List(1)) mustEqual 1
+      List.sum3(List('a')) mustEqual 'a'
+    }
+
+    "return sum of all elements in a multiple-element List" in {
+      List.sum3(List(1,2,3)) mustEqual 6
+      List.sum3(List('a','b','c')) mustEqual 'a'+'b'+'c'
+    }
+  }
+
+  "product3 (uses foldLeft)" should {
+    "result in one for an empty List" in {
+      List.product3(List()) mustEqual 1
+    }
+
+    "return the element value in a single-element List" in {
+      List.product3(List(3)) mustEqual 3
+      List.product3(List('a')) mustEqual 'a'
+    }
+
+    "return product of all elements in a multiple-element List" in {
+      List.product3(List(1,2,3)) mustEqual 6
+      List.product3(List(0,1)) mustEqual 0
+    }
+  }
+
+  "length2 (uses foldLeft)" should {
+    "result in 0 for an empty List" in {
+      List.length2(List()) mustEqual List.length(List()) 
+    }
+
+    "result in 1 for a single-element List" in {
+      List.length2(List(1)) mustEqual List.length(List(1))
+      List.length2(List('a')) mustEqual List.length(List(1))
+    }
+
+    "reslut in the number of elements in a multiple-element List" in {
+      List.length2(List(1,2,3)) mustEqual List.length(List(1,2,3))
+      List.length2(List('a','b','c')) mustEqual List.length(List('a','b','c'))
+    }
+  }
+
+  //========================================================================================
+  //#3.12
+  "reverse" should {
+    "result in Nil for an empty List" in {
+      List.reverse(List()) mustEqual Nil
+    }
+
+    "return the same list for a single-element List" in {
```

Comment:
Perfect for property-based testing: `prop { x: Int => val l = List(x); List.reverse(l) mustEqual l }`